lib/utils/nodeManager.ts

```diff
@@ -0,0 +1,422 @@
+			return false;
+		}
+		const methodInfo = this.getNodeMethodInfo(node, methodName);
+		if (!methodInfo) {
+			console.error(`NodeManager - remove breakpoint - method ${methodName} does not exist on node ${nodePath}`);
+			return false;
+		}
+		const key = `${methodInfo.className}::${methodInfo.methodName}`;
+		const keyAndPath = `${key}/${nodePath}`;
+		const breakpointInfo = this.breakpointInfoMap.get(key);
+
+		if (breakpointInfo) {
+			const breakpointData = breakpointInfo.breakpoints.get(nodePath);
+			if (breakpointData) {
+				try {
+					await RuntimeWrappers.removeBreakpointFromFunction(this.session, breakpointInfo);
+					breakpointInfo.breakpoints.delete(nodePath);
+					if (breakpointInfo.breakpoints.size === 0) {
+					// this was the only breakpoint set for the key
+						this.breakpointInfoMap.delete(key);
+						(<any>global).breakpointMethods.delete(key);
+					} else {
+						await this._setBreakpoint(
+							key, this.nodePathMap.get([...breakpointInfo.breakpoints][0][1].nodePath)!, breakpointInfo);
+					}
+					return true;
+				} catch (err) {
+					console.error('Got error removing breakpoint', err);
+					return false;
+				}
+			} else {
+				console.error(`NodeManager - remove breakpoint - did not find breakpoint for path: ${keyAndPath}`);
+				return false;
+			}
+		} else {
+			console.error(`NodeManager - remove breakpoint - did not find breakpoint id at all: ${keyAndPath}`);
+			(<any>global).breakpointMethods.delete(key);
+			return false;
+		}
+	}
+
+	// Returns the singleton instance
+	public static getInstance<S extends BlueshellState, E>(): INodeManager<S, E> {
+		if (!this.instance) {
+			this.instance = new NodeManager();
+		}
+		return this.instance;
+	}
+	public static reset() {
+		this.instance = null;
+	}
+
+	// Adds a bt node (and all its children) to be considered for debugging (setting/removing breakpoints)
+	public addNode(node: BaseNode<S, E>) {
+		const path = node.path;
+		if (this.nodePathMap.has(path)) {
+			throw new DuplicateNodeAdded(path);
+		} else {
+			this.nodePathMap.set(path, node);
+		}
+		if (isParentNode(node)) {
+			node.getChildren().forEach((child) => {
+				this.addNode(child);
+			});
+		}
+	}
+
+	// Removes a bt node (and all its children) to be considered for debugging (setting/removing breakpoints)
+	public removeNode(node: BaseNode<S, E>) {
+		if (isParentNode(node)) {
+			node.getChildren().forEach((child) => {
+				this.removeNode(child);
+			});
+		}
+		this.nodePathMap.delete(node.path);
+	}
+
+	// Given a node path, return the bt node we have cached for that path
+	public getNode(path: string): BaseNode<S, E>|undefined {
+		return this.nodePathMap.get(path);
+	}
+
+	// Shuts down the debug server
+	public async shutdown() {
+		return new Promise<void>((resolve, reject) => {
+			if (this.server) {
+				this.server.close((err) => {
+					if (err) {
+						reject(err);
+					} else {
+						resolve();
+					}
+				});
+			} else {
+				resolve();
+			}
+		});
+	}
```

Comment:
```suggestion
	public async shutdown() {
		if (this.server) {
			await util.promisify(this.server.close.bind(this.server));
		}
	}
```

I think this would work and be more concise, but I see no problem with the way it's implemented here anyway

Reply:
That is much more concise - since we're making some fixes anyways, I included this change as well (next push)